### PR TITLE
Add '--dmd' and '--dc' to print the path of the installed compiler

### PR DESCRIPTION
**Pull request by wilzbach**
This has come up frequently in the past.

Use case: one has a generic compiler variable like `$DMD` in a CI script and wants to use the install script without activating the environment.

See also:
- https://github.com/dlang/dmd/pull/7848#issuecomment-364235083
- https://github.com/dlang/tools/pull/297/files#diff-f9a62833538ab14fff849c2be16ab986R9

I'm not fully convinced about the name, so bike-shedding is allowed.

An alternative to this would have been to add a flag similar to `-a|--activate`.
That would require a two line change and would be fine for me too.

CC @WebDrake

### CODE REVIEW

File: script/install.sh [hunk 1]
```diff
@@ -150,6 +150,7 @@ Commands
   install       Install a D compiler (default command)
   uninstall     Remove an installed D compiler
   list          List all installed D compilers
+  where         Finds the binary for a D compiler (installs it if no compiler is found)
```
Comment (WebDrake) [inline script/install.sh]: That seems an odd conflation of concerns.  Asking _where_ it is shouldn't auto-install it!

File: changelog/where.dd [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+$(CONSOLE curl https://dlang.org/install.sh | bash -s where
+/home/seb/dlang/dmd-2.078.2/linux/bin64
+)
+
+$(CONSOLE ~/dlang/install.sh ldc where
+/home/seb/dlang/ldc-1.7.0/bin
+)
+
+As with `install`, if a compiler isn't found, it will be installed.
```
Comment (WebDrake) [inline changelog/where.dd]: Why conflate these two concerns?  If one issues an instruction to `install`, then clearly one is asking for something to be installed if it isn't already.

If one is asking `where` something is then it seems reasonable that one might expect the response "nowhere", rather than have it be installed just so it can definitely be found.

File: script/install.sh (hunk 1 above)
Comment (WebDrake) [reply script/install.sh]: What motivation is there behind this assumption that `where` should install if not found?

**Comment by WebDrake** [comment]
My questions seem to have got lost with the updated patch.  I'll summarize: why make `where` install a compiler if not found, rather than just report (in some appropriate way) that it is not found?

The simplest solution would probably be that `where` produces no output if the compiler is not installed, or the install location if it is.

Then, if in some script it's _wanted_ that a compiler be installed if not found, that could be implemented by querying `where` and issuing a separate `install` command if necessary.

**Comment by wilzbach** [comment]
> My questions seem to have got lost with the updated patch. 

No worries. They weren't lost. I was about to reply to them.

> I'll summarize: why make where install a compiler if not found, rather than just report (in some appropriate way) that it is not found?

The updated patch removed `where` as a separate option. It now behaves analog to `-a` | `--activate` which is only an option for `install` (the default command) and will print the path to the sourceable `activate` script.

I hppe this answers all your questions as `--where` can't be called with anything else than the `install` command?

**Comment by WebDrake** [comment]
I'm not sure it really feels like the right answer.  Could we take a pause on this and revisit tomorrow?  I need some sleep ;-)

I don't think it really satisfies what I see as _my_ use-case.  I would say that there are 2 different options that would satisfy what I'm looking for:

* provide an alternative flag for `install` that allows the installed compiler to be added to the `PATH` without overwriting all the other environment vars like `DC`, `DMD`, etc.  This feels a bit finnicky, though, because it could get tricky to track everything that's been added to the path (or not).

* provide an easy way to query the location of a named installed compiler, that would give the location if it's installed, or no output if not.

Bottom line, I _like_ the idea of the `where` command.  I just think it would satisfy my needs better if it just behaved as per the second bullet point above.  Is there a reason why that's difficult or problematic?

**Comment by wilzbach** [comment]
> I'm not sure it really feels like the right answer. Could we take a pause on this and revisit tomorrow? I need some sleep ;-)

No worries. There's quite a bottleneck for `installer` (it's only @MartinNowak who can effectively approve PRs), so the review backlog is quite long - https://github.com/dlang/installer/pulls

> I don't think it really satisfies what I see as my use-case.

I'm not sure what's your use case, but it satisfies the need for `dlang/dmd` and `dlang/tools`.

-------------------

### DMD:

Previously:

```
source ~/dlang/install.sh "$DMD"
HOST_DMD=$DMD"
deactivate
```

Now

```
HOST_DMD=$(~/dlang/install.sh --where "$DMD")
```


--- 

### tools

```
~/dlang/install.sh install ldc
LDMD2=$(find ~/dlang -type f -name "ldmd2")
```

now:

```
LDMD2=$(~/dlang/install.sh install --where ldc)
```

> provide an alternative flag for install that allows the installed compiler to be added to the PATH without overwriting all the other environment vars like DC, DMD, etc. This feels a bit finnicky, though, because it could get tricky to track everything that's been added to the path (or not).

Not a huge fan of this. It's also more complicated than this PR.

> provide an easy way to query the location of a named installed compiler, that would give the location if it's installed, or no output if not.

How would that work with `~/dlang/install install ldc`?
I mean sure you could write a `/compiler_path` file in the root directory of the ldc archive and then do a `dirname` on the output of `--activate`, but then we would break code if we ever change the location of the `activate` script.

**Comment by joseph-wakeling-sociomantic** [comment]
On thinking it over: yes, as a flag to `install` this is probably the best way of doing things.  It means that the answer is only possible at the time of install, and that's a useful simplification, because it means that there's no need to keep state hanging around, or rely on assumed rules about where compilers get placed.

A separate `where` command could work on the grounds that the install location for a given specified compiler is 100% deterministic (for the install script) based on the compiler name.  But of course this is not quite true when one considers generic `ldc`, `dmd`, `gdc` as opposed to specified versions.  So best avoided.

Something that's not clear from your above examples: would the result be the _directory_ where the compiler lives, or the full path to the compiler executable?  I ask because:

```
LDMD2=$(find ~/dlang -type f -name "ldmd2")
```

versus

```
LDMD2=$(~/dlang/install.sh install --where ldc)
```

... do not seem implicitly to be the same thing.  After all, with LDC and GDC there are _two_ compiler executables installed, with 2 different names (`ldc2` and `ldmd2`, and `gdc` and `gdmd`).

File: script/install.sh [hunk 3]
```diff
@@ -192,6 +193,7 @@ Description
 Options
 
   -a --activate     Only print the path to the activate script
+  -w --where        Only print the path to the compiler binary
```
Comment (WebDrake) [inline script/install.sh]: If I understand this code correctly, it's not the path to the compiler binary.  It's the path to the directory where the compiler binary has been installed.

File: script/install.sh [hunk 4]
```diff
@@ -313,43 +328,25 @@ parse_args() {
...
+                    local suffix=.fish
+                fi
+                if [ "$VERBOSITY" -eq 0 ]; then
+                    echo "$ROOT/$2/activate${suffix:-}"
                 else
-                    log "Using dub shipped with $2"
+                    log "
+Run \`source     $ROOT/$2/activate${suffix:-}\` in your shell to use $2.
+This will se    tup PATH, LIBRARY_PATH, LD_LIBRARY_PATH, DMD, DC, and PS1.
+Run \`deacti    vate\` later on to restore your environment."
```
Comment (WebDrake) [inline script/install.sh]: What's the reason for the weird whitespace on these lines?

File: changelog/where.dd [hunk 5]
```diff
@@ -0,0 +1,15 @@
+`~/dlang/install.sh` now supports `--where`
+
+`--where` has been added as a new option to the install script.
+It offers a convenient way to get the path of a specific compiler.
```
Comment (WebDrake) [inline changelog/where.dd]: `path of a specific compiler` is a bit vague -- `--where` doesn't provide the path to the compiler, it provides the directory where the compiler has been installed.

**Comment by wilzbach** [comment]
> Something that's not clear from your above examples: would the result be the directory where the compiler lives, or the full path to the compiler executable? I ask because:
> ...
> ... do not seem implicitly to be the same thing. After all, with LDC and GDC there are two compiler executables installed, with 2 different names (ldc2 and ldmd2, and gdc and gdmd).

Contrary to the previous changelog entry, they were doing the same thing (`--where` was returning `<path-to-ldc-bin>/ldmd2`). Sorry for the confusion!

However, after thinking about this, I think I found an even better solution with `--dmd` and `--dc`.
This is analog to the existing `$DMD` and `$DC` variables that _this_ install script exposes.
I like this more because it's explicit what interface the user wants and no `dirname` hacks need to be done to switch the interface of the compiler from `ldmd` to `ldc`.

File: changelog/install-dmd-dc.dd [hunk 6]
```diff
@@ -0,0 +1,27 @@
+`~/dlang/install.sh` now supports `--dmd` and `--dc`
+
+`--dmd` and `--dc` has been added as a new option to the install script.
+These flags work analogous to the existing `--activate` flag and
+print information about the installed compiler binary:
+
+$(UL
+    $(LI `--dmd` returns the path of a specific compiler executable (DMD-like interface))
+    $(LI `--dc` returns the path of a specific compiler executable (custom compiler interface))
```
Comment (joseph-wakeling-sociomantic) [inline changelog/install-dmd-dc.dd]: Can this be the same as what `--dmd` gives?  Might be worth noting if so.

File: changelog/install-dmd-dc.dd [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+$(H4 Examples)
+
+$(CONSOLE curl https://dlang.org/install.sh | bash -s --dmd
+/home/seb/dlang/dmd-2.078.2/linux/bin64/dmd
```
Comment (joseph-wakeling-sociomantic) [inline changelog/install-dmd-dc.dd]: What would `--dc` give you if the compiler is `dmd`?

File: changelog/install-dmd-dc.dd [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+print information about the installed compiler binary:
+
+$(UL
+    $(LI `--dmd` returns the path of a specific compiler executable (DMD-like interface))
+    $(LI `--dc` returns the path of a specific compiler executable (custom compiler interface))
+)
+
+$(H4 Examples)
+
+$(CONSOLE curl https://dlang.org/install.sh | bash -s --dmd
```
Comment (joseph-wakeling-sociomantic) [inline changelog/install-dmd-dc.dd]: Not sure it's really helpful to put in this `curl` stuff.  It distracts from the example of how to use the `install.sh` script itself.

File: changelog/install-dmd-dc.dd [hunk 9]
```diff
@@ -0,0 +1,27 @@
+`~/dlang/install.sh` now supports `--dmd` and `--dc`
+
+`--dmd` and `--dc` has been added as a new option to the install script.
+These flags work analogous to the existing `--activate` flag and
+print information about the installed compiler binary:
+
+$(UL
+    $(LI `--dmd` returns the path of a specific compiler executable (DMD-like interface))
```
Comment (joseph-wakeling-sociomantic) [inline changelog/install-dmd-dc.dd]: What will this be if the compiler does not expose a DMD-like interface?

File: changelog/install-dmd-dc.dd (hunk 8 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd]: It's how you get the installer script and would use it for a CI script.
It also provides a seamless experience for readers who want to try out the new flags as they can just copy/paste the command.

File: changelog/install-dmd-dc.dd (hunk 7 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd]: The same (see the tests). The changelog isn't intended to be a full documentation, but to raise awareness of new features or improvements. dlang.org/install.html will come with more extensive documentation.

File: changelog/install-dmd-dc.dd (hunk 9 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd]: I don't think SDC will grow into a full compiler any time soon and even if, for a seamless experience it most likely will implement the DMD interface (IIRC it already does so).
The same applies for other compilers and if there's ever a case in a far far far distant future where we would want to include a new compiler, but are too lazy to wrap the DMD interface around it, `fatal` error is the Unix way.
Also as mentioned this is the changelog entry - not the documentation, so imho there's no need to go into details here.

Comment (WebDrake) [reply changelog/install-dmd-dc.dd]: > Also as mentioned this is the changelog entry - not the documentation, so imho there's no need to go into details here.

I didn't ask for the changelog to be altered.  I asked what the result will be if the compiler does not expose a DMD-like interface.

File: changelog/install-dmd-dc.dd (hunk 8 above)
Comment (WebDrake) [reply changelog/install-dmd-dc.dd]: Sure, but it would be clearer if the example was annotated to that effect, e.g. `download and run install script, printing out the installed compiler`.

File: script/install.sh [hunk 10]
```diff
@@ -192,6 +195,8 @@ Description
 Options
 
   -a --activate     Only print the path to the activate script
+  --dmd             Only print the path to the compiler binary (e.g. dmd, ldmd2, gdmd)
+  --dc              Only print the path to the compiler binary (e.g. dmd, ldc, gdc)
```
Comment (WebDrake) [inline script/install.sh]: * What will happen if I use both `--dmd` and `--dc` together?

* It looks like this documentation could be clearer about the distinction between the flags, i.e. that one is for the DMD-alike compiler binary

**Comment by MartinNowak** [comment]
> HOST_DMD=$(~/dlang/install.sh --where "$DMD")

Sorry, I've only grokked part of the discussion.

Just using the compiler without the environment is not reliable, as the sc.ini/dmd.conf in your user home take precendence over the ones next to the compiler binary :/.
I'm still convinced that a lot of non-sense is rooted on this single problem. Why would a makefile need an absolute path to the compiler? Just so that it can for usage of `-conf=` or `-conf=path/to/dmd.conf`.
https://github.com/dlang/dmd/pull/4256

**Comment by wilzbach** [comment]
> Just using the compiler without the environment is not reliable, as the sc.ini/dmd.conf in your user home take precendence over the ones next to the compiler binary :/.
> I'm still convinced that a lot of non-sense is rooted on this single problem. 

Urgh I wasn't even aware of this. Let's fix this mess then -> https://github.com/dlang/dmd/pull/7915

> Sorry, I've only grokked part of the discussion.

The idea is to have a `--dmd` and `--dc` flag which returns the location of the compiler binary (they return the same output as `$DMD` and `$DC`), s.t
- the environment doesn't need to be loaded which turns out is very error-prone (e.g. our recent CI fiasco at dmd)
- multiple compiler can be installed and used in parallel without needing to switch environments (in @WebDrake's use case for dlang/tools the rdmd testsuite even accepts multiple compiler binaries)

> Just using the compiler without the environment is not reliable, as the sc.ini/dmd.conf in your user home take precendence over the ones next to the compiler binary :

Well, that wouldn't be a huge problem for CIs (the main motivation for this feature), but of course one could always do sth. like this with `dirname`:

```shell
DMD=$(~/dlang/install.sh 2.078.3  --dmd)
make DMD="$DMD" DFLAGS="-conf="$(dirname "$DMD")/dmd.conf)"
```

In any case the motivation was that `--dmd` is cleaner and less prone to errors than `find ~/dlang -type f -name "ldmd2"` [which is currently used](https://github.com/dlang/tools/blob/master/travis.sh).

**Comment by WebDrake** [comment]
@MartinNowak a specific example where multiple compilers are wanted comes from here:
https://github.com/dlang/tools/blob/master/travis.sh#L5-L16

The goal there is to have one build of `rdmd` tested with multiple different compilers passed via the `--compiler` flag.

Now that said, I'm not wedded to that particular way of doing things: it was a step forward given the existing test setup, but there are multiple things that probably can and should be changed.

Note that the workaround I implemented for that use case (i.e. using `find`) seems to be working OK for the time being, so from my side there is no urgency to this `--where` flag or anything like it.

File: changelog/where.dd (hunk 2 above)
Comment (CyberShadow) [reply changelog/where.dd]: I think installation should be available as a flag for all commands which imply using a specific installer version.

I.e. `source ~/dlang/install.sh activate -i 2.080.0` should install 2.080.0 if it's not installed, then activate it.

File: changelog/install-dmd-dc.dd [hunk 11]
```diff
@@ -0,0 +1,27 @@
+`~/dlang/install.sh` now supports `--dmd` and `--dc`
```
Comment (CyberShadow) [inline changelog/install-dmd-dc.dd]: These are not options, they're actions. They should be named "get-dmd-path" and "get-dc-path" OSLT.

Am I missing something?

File: script/install.sh (hunk 10 above)
Comment (CyberShadow) [reply script/install.sh]: Yep, conceptually these are actions, not flags.

**Comment by AndrewEdwards** [comment]
@wilzbach @WebDrake @joseph-wakeling-sociomantic @CyberShadow @MartinNowak 

What is the current stand on this? Is there actually something preventing this from being merged? With the exception of the requested change from "--dmd" and "--dc" to "get-dmd-path" and "get-dc-path" all other concerns appears to have been addressed. @wilzbach can you address @CyberShadow's concern and resolve the conflicts to prepare this for merger? @MartinNowak, what other concerns if any do you have about this pull request?

File: changelog/install-dmd-dc.dd (hunk 9 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd:8]: This can't happen as `install.sh` only supports GDC, DMD and SDC. If there's ever a new compiler which would be added to `install.sh`, it would fail, I presume.

File: changelog/install-dmd-dc.dd (hunk 6 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd:9]: I added an example.

File: changelog/install-dmd-dc.dd (hunk 7 above)
Comment (wilzbach) [reply changelog/install-dmd-dc.dd]: > What would --dc give you if the compiler is dmd?

I added an example + it's not the documentation, just a changelog.

**Comment by wilzbach** [comment]
> @wilzbach can you address @CyberShadow's concern and resolve the conflicts to prepare this for merger? 

Ok. I reworked it. I combined them under a single `get-path` action and added support for getting the path to the requested dub binary as well.

**Comment by CyberShadow** [comment]
Is it possible to remove those annoying CodeCov annotations? They make the diff unreadable.

**Comment by AndrewEdwards** [comment]
I see what you mean. Sorry, not sure how to do that. I'll defer to @wilzbach.

**Comment by CyberShadow** [comment]
```js
for (var n of document.querySelectorAll('.check-annotation')) n.parentElement.remove(n);
```
in the console does it.

File: test/common.sh [hunk 12]
```diff
@@ -4,3 +4,19 @@ set -eu -o pipefail
...
 INSTALLER="$ROOT/script/install.sh"
+
+assert() {
+    actual="$1"
+    expected="$2"
+    if [ "$actual" != "$expected" ] ; then
+        echo "Actual: $actual"
+        echo "Expected: $expected"
+        exit 1
+    fi
```
Comment (CyberShadow) [inline test/common.sh:15]: For this I would simply write `diff -u <(cat <<< "$1") <(cat <<< "$2")`, but this is OK too.

**Comment by AndrewEdwards** [comment]
I'll need to set up my environment first. Currently just using the web interface. As a temporary fix, you can disable them per file by unchecking the show annotations option.

disable-annotations

**Comment by CyberShadow** [comment]
Ah, thanks. I missed that. That's much simpler.

**Review by CyberShadow — APPROVED** [review]
Thank you. Good work on the thorough test script.

**Comment by wilzbach** [comment]
Wow. I didn't think this would ever get merged. Thanks!

I opened a documentation PR -> https://github.com/dlang/dlang.org/pull/2849
